pmap/rpc_prog.go

```diff
@@ -0,0 +1,159 @@
+package pmap
+
+import (
+	"github.com/prashanthpai/sunrpc"
+)
+
+const (
+	portmapProgNum     = 34123456 // GLUSTER_PMAP_PROGRAM
+	portmapProgVersion = 1        // GLUSTER_PMAP_VERSION
+)
+
+const (
+	gfPmapNull        = iota
+	gfPmapPortByBrick // GF_PMAP_PORTBYBRICK
+	gfPmapBrickByPort // GF_PMAP_BRICKBYPORT
+	gfPmapSignUp      // Don't use
+	gfPmapSignIn      // GF_PMAP_SIGNIN
+	gfPmapSignOut     // GF_PMAP_SIGNOUT
+)
+
+// GfPortmap is a type for GlusterFS Portmap RPC program
+type GfPortmap struct {
+	name        string
+	progNum     uint32
+	progVersion uint32
+	procedures  []sunrpc.Procedure
+}
+
+// NewGfPortmap returns a new instance of GfPortmap type
+func NewGfPortmap() *GfPortmap {
+	// rpc/rpc-lib/src/protocol-common.h
+	return &GfPortmap{
+		name:        "Gluster Portmap",
+		progNum:     portmapProgNum,
+		progVersion: portmapProgVersion,
+		procedures: []sunrpc.Procedure{
+			sunrpc.Procedure{
+				sunrpc.ProcedureID{ProgramNumber: portmapProgNum, ProgramVersion: portmapProgVersion,
+					ProcedureNumber: gfPmapPortByBrick}, "PortByBrick"},
+			sunrpc.Procedure{
+				sunrpc.ProcedureID{ProgramNumber: portmapProgNum, ProgramVersion: portmapProgVersion,
+					ProcedureNumber: gfPmapBrickByPort}, "BrickByPort"},
+			sunrpc.Procedure{
+				sunrpc.ProcedureID{ProgramNumber: portmapProgNum, ProgramVersion: portmapProgVersion,
+					ProcedureNumber: gfPmapSignIn}, "SignIn"},
+			sunrpc.Procedure{
+				sunrpc.ProcedureID{ProgramNumber: portmapProgNum, ProgramVersion: portmapProgVersion,
+					ProcedureNumber: gfPmapSignOut}, "SignOut"},
+		},
+	}
+}
+
+// Name returns the name of the RPC program
+func (p *GfPortmap) Name() string {
```

Comment:
An interface is just a set of functions, that any type can implement. This is unlike classes in other languages, where the class itself is a type as well as associated functions. So as it is, we cannot directly use the objects returned here.

But looking at the rest of the code, it seems we've been using the same struct everywhere. So we could just define a common struct for a Program and simplify a lot of stuff.

@prashanthpai Can you please do this?